gcloud/datastore/helpers.py

```diff
@@ -155,6 +155,48 @@ def entity_from_protobuf(pb):
     return entity
 
 
+def _set_pb_meaning_from_entity(entity, name, value, value_pb,
+                                is_list=False):
+    """Add meaning information (from an entity) to a protobuf.
+
+    :type entity: :class:`gcloud.datastore.entity.Entity`
+    :param entity: The entity to be turned into a protobuf.
+
+    :type name: string
+    :param name: The name of the property.
+
+    :type value: object
+    :param value: The current value stored as property ``name``.
+
+    :type value_pb: :class:`gcloud.datastore._generated.entity_pb2.Value`
+    :param value_pb: The protobuf value to add meaning / meanings to.
+
+    :type is_list: bool
+    :param is_list: (Optional) Boolean indicating if the ``value`` is
+                    a list value.
+    """
+    if name not in entity._meanings:
+        return
+
+    meaning, orig_value = entity._meanings[name]
+    # Only add the meaning back to the protobuf if the value is
+    # unchanged from when it was originally read from the API.
+    if orig_value is not value:
+        return
+
+    # For lists, we set meaning on each sub-element.
+    if is_list:
```

Comment:
Gotcha. The reasoning behind why it was `[A, B]` or nothing was that meanings were completely foreign outside of GAE so modifying a property that had a meaning would drop all that information.

We could track individual elements, but it is almost certainly more trouble than it's worth. The check `orig_value is not value` means literally the same object. A `list` is mutable though, so users could make it shorter or longer or move items into a different order without making it a different object. Hence our decision to just "assume" that no changes means no changes.